lib/util/source-code.js

```diff
@@ -223,14 +224,110 @@ class SourceCode extends TokenStore {
     }
 
     /**
+     * Gets all comments before the given node or token.
+     * @param {ASTNode|token} nodeOrToken The AST node or token to get the comments for.
+     * @returns {Object} The list of comments indexed by their position.
+     * @public
+     */
+    getCommentsBefore(nodeOrToken) {
+        if (this._commentCache.has("getCommentsBefore", nodeOrToken)) {
+            return this._commentCache.get("getCommentsBefore", nodeOrToken);
+        }
+
+        const comments = [];
+
+        /**
+         * The Program node's range[0] corresponds to the start value of the
+         * first node (ignoring comment tokens at the beginning of the program).
+         * As a result, calling this.getTokenBefore() with the Program node
+         * will return comments at the beginning of a file.
+         */
+        if (nodeOrToken.type === "Program") {
+            return comments;
+        }
+
+        let currentToken = this.getTokenBefore(nodeOrToken, { includeComments: true });
+
+        while (currentToken && astUtils.isCommentToken(currentToken)) {
+            comments.unshift(currentToken);
+            currentToken = this.getTokenBefore(currentToken, { includeComments: true });
+        }
+
+        this._commentCache.set("getCommentsBefore", nodeOrToken, comments);
+        return comments;
+    }
+
+    /**
+     * Gets all comments after the given node or token.
+     * @param {ASTNode|token} nodeOrToken The AST node or token to get the comments for.
+     * @returns {Object} The list of comments indexed by their position.
+     * @public
+     */
+    getCommentsAfter(nodeOrToken) {
+        if (this._commentCache.has("getCommentsAfter", nodeOrToken)) {
+            return this._commentCache.get("getCommentsAfter", nodeOrToken);
+        }
+
+        const comments = [];
+
+        /**
+         * The Program node's range[1] corresponds to the end value of the
+         * last node (ignoring comment tokens at the end of the program).
+         * As a result, calling this.getTokenAfter() with the Program node
+         * will return comments at the end of a file.
+         */
+        if (nodeOrToken.type === "Program") {
+            return comments;
```

Comment:
My main concern is with consistency. The following two should yield the same results:
```
// comment
/* another comment */
```
and
```
// comment
/* another comment */
var foo;
```

Without some kind of forcing function, they'll actually be treated differently. In my mind, the Program node represents the entire file - not just the executable code. What's really confusing to me is the Program node having mismatching `start` and `range[0]` values, make it unclear what the intention is at the parser level.

I wonder if this is intentional and I'm just missing something or if it's a bug? I guess I can file a bug report on Espree either way.